File: backend/services/digest.py
"""Weekly digest email (Monday 09:00 Sydney)."""
from datetime import datetime, timezone, timedelta

from deps import (
    leads_collection, share_events_collection,
    RESEND_API_KEY, RESEND_FROM_EMAIL, ADMIN_NOTIFICATION_EMAILS,
    logger,
)


async def build_weekly_digest() -> dict:
    now = datetime.now(timezone.utc)
    week_ago = (now - timedelta(days=7)).isoformat()
    new_leads_count = await leads_collection.count_documents({"created_at": {"$gte": week_ago}})
    pipeline_row = leads_collection.aggregate([
        {"$match": {"created_at": {"$gte": week_ago}}},
        {"$group": {"_id": None, "total": {"$sum": "$estimated_refund"}}},
    ])
    new_pipeline = 0.0
    async for r in pipeline_row:
        new_pipeline = float(r.get("total", 0))

    channels = ["download", "native", "copy", "story_download"]
    channel_counts = {}
    for ch in channels:
        channel_counts[ch] = await share_events_collection.count_documents(
            {"channel": ch, "created_at": {"$gte": week_ago}}
        )
    top_channel = max(channel_counts.items(), key=lambda kv: kv[1]) if channel_counts else ("-", 0)

    top_cursor = leads_collection.aggregate([
        {"$match": {"referred_by_lead_id": {"$ne": None}, "created_at": {"$gte": week_ago}}},
        {"$group": {"_id": "$referred_by_lead_id", "count": {"$sum": 1}}},
        {"$sort": {"count": -1}},
        {"$limit": 3},
    ])
    top_referrers = []
    async for row in top_cursor:
        ref = await leads_collection.find_one(
            {"id": row["_id"]},
            {"_id": 0, "first_name": 1, "email": 1, "referral_code": 1},
        )
        if ref:
            top_referrers.append({**ref, "referred_count": row["count"]})

    return {
        "since": week_ago,
        "new_leads_count": new_leads_count,
        "new_pipeline_value": round(new_pipeline, 2),
        "share_events_by_channel": channel_counts,
        "top_channel": {"channel": top_channel[0], "count": top_channel[1]},
        "top_referrers": top_referrers,
    }


def _digest_to_html(d: dict) -> str:
    rows = "".join(
        f"<li><strong>{r['first_name']}</strong> ({r.get('email', '')}) — "
        f"code <code>{r.get('referral_code', '')}</code> — "
        f"{r['referred_count']} new referrals</li>"
        for r in d["top_referrers"]
    ) or "<li>No referred leads this week.</li>"
    return f"""
    <h2>refundmysuper — weekly digest</h2>
    <p>Since {d['since'][:10]}</p>
    <ul>
      <li><strong>{d['new_leads_count']}</strong> new leads</li>
      <li><strong>${d['new_pipeline_value']:,.0f}</strong> new pipeline value</li>
      <li>Top share channel: <strong>{d['top_channel']['channel']}</strong>
          ({d['top_channel']['count']} events)</li>
    </ul>
    <h3>Top referrers this week</h3>
    <ol>{rows}</ol>
    <p style="color:#4A5D68;font-size:12px;">This digest was generated automatically. Configure ADMIN_NOTIFICATION_EMAILS + RESEND_API_KEY to deliver it.</p>
    """


async def send_weekly_digest() -> dict:
    digest = await build_weekly_digest()
    html = _digest_to_html(digest)
    admin_recipients = [e.strip() for e in ADMIN_NOTIFICATION_EMAILS.split(",") if e.strip()]
    if RESEND_API_KEY and RESEND_FROM_EMAIL and admin_recipients:
        try:
            import resend
            resend.api_key = RESEND_API_KEY
            resend.Emails.send({
                "from": RESEND_FROM_EMAIL,
                "to": admin_recipients,
                "subject": f"refundmysuper weekly digest — {digest['new_leads_count']} new leads",
                "html": html,
            })
            logger.info("Weekly digest sent to %s", admin_recipients)
        except Exception as e:
            logger.exception("Weekly digest send failed: %s", e)
    else:
        logger.info(
            "[STUB] Weekly digest not sent — RESEND / ADMIN_NOTIFICATION_EMAILS not configured. "
            "%d new leads · $%s pipeline · top channel=%s",
            digest["new_leads_count"],
            f"{digest['new_pipeline_value']:,.0f}",
            digest["top_channel"]["channel"],
        )
    return digest

File: src/lib/server/digest.js
// Weekly digest email. Port of backend/services/digest.py. Cron scheduling
// (Mon 09:00 Sydney) is explicitly deferred — this is invoked on-demand via
// the admin UI's "Run digest" button / POST /api/admin/weekly-digest/run.
import { getSupabaseAdmin } from "./supabaseAdmin";
import { sendMail } from "./mailer";
import { renderEmail, statBlock, button, escapeHtml, BRAND, SITE_URL } from "./emailTemplates";

const CHANNEL_LABEL = { download: "Downloads", native: "Native share", copy: "Link copies", story_download: "Story downloads" };

export async function buildWeeklyDigest() {
  const supabase = getSupabaseAdmin();
  const weekAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000).toISOString();

  const { count: newLeadsCount } = await supabase
    .from("leads")
    .select("*", { count: "exact", head: true })
    .gte("created_at", weekAgo);

  const { data: pipelineRows } = await supabase.from("leads").select("estimated_refund").gte("created_at", weekAgo);
  const newPipeline = (pipelineRows || []).reduce((sum, r) => sum + Number(r.estimated_refund || 0), 0);

  const channels = ["download", "native", "copy", "story_download"];
  const channelCounts = {};
  for (const ch of channels) {
    const { count } = await supabase
      .from("share_events")
      .select("*", { count: "exact", head: true })
      .eq("channel", ch)
      .gte("created_at", weekAgo);
    channelCounts[ch] = count || 0;
  }
  const topChannelEntry = Object.entries(channelCounts).sort((a, b) => b[1] - a[1])[0] || ["-", 0];

  const { data: referredRows } = await supabase
    .from("leads")
    .select("referred_by_lead_id")
    .not("referred_by_lead_id", "is", null)
    .gte("created_at", weekAgo);
  const countsByReferrer = {};
  for (const row of referredRows || []) {
    countsByReferrer[row.referred_by_lead_id] = (countsByReferrer[row.referred_by_lead_id] || 0) + 1;
  }
  const topEntries = Object.entries(countsByReferrer)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 3);
  const topReferrers = [];
  for (const [leadId, count] of topEntries) {
    const { data: ref } = await supabase
      .from("leads")
      .select("first_name, email, referral_code")
      .eq("id", leadId)
      .maybeSingle();
    if (ref) topReferrers.push({ ...ref, referred_count: count });
  }

  return {
    since: weekAgo,
    new_leads_count: newLeadsCount || 0,
    new_pipeline_value: Math.round(newPipeline * 100) / 100,
    share_events_by_channel: channelCounts,
    top_channel: { channel: topChannelEntry[0], count: topChannelEntry[1] },
    top_referrers: topReferrers,
  };
}

function digestToHtml(d) {
  const referrerRows = d.top_referrers.length
    ? d.top_referrers
        .map(
          (r, i) => `
        <tr>
          <td style="padding:12px 16px;background:${i % 2 === 0 ? "#F8FAFC" : "#FFFFFF"};font-family:Arial,Helvetica,sans-serif;font-size:13px;color:${BRAND.ink};border-bottom:1px solid ${BRAND.border};">
            <strong>${escapeHtml(r.first_name)}</strong><br/>
            <span style="color:${BRAND.muted};font-size:12px;">${escapeHtml(r.email || "")}</span>
          </td>
          <td style="padding:12px 16px;background:${i % 2 === 0 ? "#F8FAFC" : "#FFFFFF"};font-family:Arial,Helvetica,sans-serif;font-size:13px;color:${BRAND.navy};border-bottom:1px solid ${BRAND.border};">
            <code style="background:#EBF3FA;padding:2px 6px;border-radius:6px;">${escapeHtml(r.referral_code || "")}</code>
          </td>
          <td align="right" style="padding:12px 16px;background:${i % 2 === 0 ? "#F8FAFC" : "#FFFFFF"};font-family:Arial,Helvetica,sans-serif;font-size:13px;font-weight:700;color:${BRAND.ink};border-bottom:1px solid ${BRAND.border};">
            ${r.referred_count}
          </td>
        </tr>`
        )
        .join("")
    : `<tr><td colspan="3" style="padding:16px;text-align:center;font-family:Arial,Helvetica,sans-serif;font-size:13px;color:${BRAND.muted};">No referred leads this week.</td></tr>`;

  const topChannelLabel = CHANNEL_LABEL[d.top_channel.channel] || d.top_channel.channel;

  return renderEmail({
    preheader: `${d.new_leads_count} new leads · $${d.new_pipeline_value.toLocaleString()} new pipeline this week.`,
    bodyHtml: `
      <h1 style="margin:0 0 4px;font-family:Arial,Helvetica,sans-serif;font-size:22px;font-weight:700;color:${BRAND.navy};">Weekly digest</h1>
      <p style="margin:0 0 24px;font-size:13px;color:${BRAND.muted};">Since ${escapeHtml(d.since.slice(0, 10))}</p>

      <table role="presentation" width="100%" cellpadding="0" cellspacing="0" style="margin-bottom:28px;">
        <tr>
          ${statBlock({ label: "New leads", value: d.new_leads_count })}
          ${statBlock({ label: "New pipeline", value: `$${d.new_pipeline_value.toLocaleString()}`, accent: BRAND.gold })}
          ${statBlock({ label: "Top channel", value: topChannelLabel, accent: BRAND.blue })}
        </tr>
      </table>

      <h2 style="margin:0 0 12px;font-family:Arial,Helvetica,sans-serif;font-size:16px;color:${BRAND.navy};">Top referrers this week</h2>
      <table role="presentation" width="100%" cellpadding="0" cellspacing="0" style="border:1px solid ${BRAND.border};border-radius:12px;overflow:hidden;margin-bottom:28px;">
        ${referrerRows}
      </table>

      ${button({ href: `${SITE_URL}/admin`, label: "Open admin dashboard" })}
    `,
  });
}

export async function sendWeeklyDigest() {
  const digest = await buildWeeklyDigest();
  const html = digestToHtml(digest);
  const adminRecipients = (process.env.ADMIN_NOTIFICATION_EMAILS || "")
    .split(",")
    .map((e) => e.trim())
    .filter(Boolean);
  if (process.env.SMTP_HOST && adminRecipients.length) {
    await sendMail({
      to: adminRecipients,
      subject: `refundmysuper weekly digest — ${digest.new_leads_count} new leads`,
      html,
    });
    console.log(`Weekly digest sent to ${adminRecipients}`);
  } else {
    console.log(
      `[STUB] Weekly digest not sent — SMTP / ADMIN_NOTIFICATION_EMAILS not configured. ` +
        `${digest.new_leads_count} new leads · $${digest.new_pipeline_value.toLocaleString()} pipeline · top channel=${digest.top_channel.channel}`
    );
  }
  return digest;
}
